klippy/extras/input_shaper.py

```diff
@@ -0,0 +1,134 @@
+                , 'zvd': ffi_lib.INPUT_SHAPER_ZVD
+                , 'zvdd': ffi_lib.INPUT_SHAPER_ZVDD
+                , 'zvddd': ffi_lib.INPUT_SHAPER_ZVDDD
+                , 'mzv': ffi_lib.INPUT_SHAPER_MZV
+                , 'ei': ffi_lib.INPUT_SHAPER_EI
+                , '2hump_ei': ffi_lib.INPUT_SHAPER_2HUMP_EI
+                , '3hump_ei': ffi_lib.INPUT_SHAPER_3HUMP_EI}
+        shaper_type = config.getchoice('shaper_type', self.shapers, None)
+        if shaper_type is None:
+            self.shaper_type_x = config.getchoice(
+                    'shaper_type_x', self.shapers, 'mzv')
+            self.shaper_type_y = config.getchoice(
+                    'shaper_type_y', self.shapers, 'mzv')
+        else:
+            self.shaper_type_x = self.shaper_type_y = shaper_type
+        self.stepper_kinematics = []
+        self.orig_stepper_kinematics = []
+        # Register gcode commands
+        gcode = self.printer.lookup_object('gcode')
+        gcode.register_command("SET_INPUT_SHAPER",
+                               self.cmd_SET_INPUT_SHAPER,
+                               desc=self.cmd_SET_INPUT_SHAPER_help)
+    def connect(self):
+        self.toolhead = self.printer.lookup_object("toolhead")
+        kin = self.toolhead.get_kinematics()
+        # Lookup stepper kinematics
+        ffi_main, ffi_lib = chelper.get_ffi()
+        steppers = kin.get_steppers()
+        for s in steppers:
+            sk = ffi_main.gc(ffi_lib.input_shaper_alloc(), ffi_lib.free)
+            orig_sk = s.set_stepper_kinematics(sk)
+            res = ffi_lib.input_shaper_set_sk(sk, orig_sk)
+            if res < 0:
+                s.set_stepper_kinematics(orig_sk)
+                continue
+            s.set_trapq(self.toolhead.get_trapq())
+            self.stepper_kinematics.append(sk)
+            self.orig_stepper_kinematics.append(orig_sk)
+        # Configure initial values
+        self.old_delay = 0.
+        self._set_input_shaper(self.shaper_type_x, self.shaper_type_y,
+                               self.shaper_freq_x, self.shaper_freq_y,
+                               self.damping_ratio_x, self.damping_ratio_y)
+    def _set_input_shaper(self, shaper_type_x, shaper_type_y
+                          , shaper_freq_x, shaper_freq_y
+                          , damping_ratio_x, damping_ratio_y):
+        if (shaper_type_x != self.shaper_type_x
+                or shaper_type_y != self.shaper_type_y):
+            self.toolhead.flush_step_generation()
+        ffi_main, ffi_lib = chelper.get_ffi()
+        new_delay = max(
+                ffi_lib.input_shaper_get_step_generation_window(
+                    shaper_type_x, shaper_freq_x, damping_ratio_x),
+                ffi_lib.input_shaper_get_step_generation_window(
+                    shaper_type_y, shaper_freq_y, damping_ratio_y))
+        self.toolhead.note_step_generation_scan_time(new_delay,
+                                                     old_delay=self.old_delay)
```

Comment:
`self.old_delay` needs to be updated afterwards.

Reply:
Good catch!